pkg/auth/audit/audit.go

```diff
@@ -148,31 +164,72 @@ func (a *auditLog) write(userInfo *User, reqHeaders, resHeaders http.Header, res
 	}
 
 	var buffer bytes.Buffer
+
 	alByte, err := json.Marshal(a.log)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to marashal log message: %w", err)
 	}
 
 	buffer.Write(bytes.TrimSuffix(alByte, []byte("}")))
-	if a.writer.Level >= levelRequest && len(a.reqBody) > 0 {
-		buffer.WriteString(`,"requestBody":`)
-		buffer.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, a.reqBody), []byte("\n")))
-	}
-	if a.writer.Level >= levelRequestResponse && resHeaders.Get("Content-Type") == contentTypeJSON && len(resBody) > 0 {
-		buffer.WriteString(`,"responseBody":`)
-		buffer.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, resBody), []byte("\n")))
+	a.writeRequest(&buffer)
+
+	if err = a.writeResponse(&buffer, resHeaders, resBody); err != nil {
+		return err
 	}
+
 	buffer.WriteString("}")
 
 	var compactBuffer bytes.Buffer
 	err = json.Compact(&compactBuffer, buffer.Bytes())
 	if err != nil {
-		return errors.Wrap(err, "compact audit log json failed")
+		return fmt.Errorf("compact audit log json failed %w", err)
 	}
 
 	compactBuffer.WriteString("\n")
+
 	_, err = a.writer.Output.Write(compactBuffer.Bytes())
-	return err
+	if err != nil {
+		return fmt.Errorf("failed to write log to output: %w", err)
+	}
+
+	return nil
+}
+
+// writeRequest attempts to write the API request to the log message.
+func (a *auditLog) writeRequest(buf *bytes.Buffer) {
+	if a.writer.Level < LevelRequest || len(a.reqBody) == 0 {
+		return
+	}
+
+	buf.WriteString(`,"requestBody":`)
+	buf.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, a.reqBody), []byte("\n")))
+}
+
+// writeResponse attempt to write the API response to the log message.
+func (a *auditLog) writeResponse(buf *bytes.Buffer, resHeaders http.Header, resBody []byte) (err error) {
+	if a.writer.Level < LevelRequestResponse || resHeaders.Get("Content-Type") != contentTypeJSON || len(resBody) == 0 {
+		return nil
+	}
+
+	switch resHeaders.Get("Content-Encoding") {
+	case contentEncodingGZIP:
+		resBody, err = decompressGZIP(resBody)
+	case contentEncodingZLib:
+		resBody, err = decompressZLip(resBody)
```

Comment:
When is this type of compression used? I see usage of gzip but not zlip in the k8s source code

Reply:
I didn't see it in the source code, but the request header has `"Accept-Encoding":["gzip, deflate, br"]`. Which means are client should be able to handle it as a valid `Content-Encoding`. Since `zlib` is in the standard library, implementing it along with `gzip` is trivial and protects us from running into the same issue if someone decides to use the `deflate` encoding. Though `br` is also in the list I chose not to add code to handle `Brotli` since it is not widely used or in the standard library. [Golang Brotli proposal ](https://github.com/golang/go/issues/44444).